Make the simplifier preserve syntactically equal types [blocks: #2064]

tautschnig opened the pull request:
base_type_eq guarantees semantic equality, but the resulting expression should
always have exactly the same type.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening commented on src/util/simplify_expr_int.cpp:
```diff
@@ -955,8 +955,13 @@ bool simplify_exprt::simplify_concatenation(exprt &expr)
+    {
+      tmp.make_typecast(expr.type());
+      simplify_node(tmp);
+    }
```
This one is potentially problematic, as the typecast is semantic, and does not necessarily preserve the bit-pattern.
I would refuse to do the simplification if the type differs.

tautschnig replied:
Ok, I’ll change this as suggested.

kroening commented:
Note that this may create further type-inconsistency in the expression tree; say a symbol expression might now have a type that is not syntactically identical to the type it has elsewhere or in the symbol table.

tautschnig commented:
> Note that this may create further type-inconsistency in the expression tree; say a symbol expression might now have a type that is not syntactically identical to the type it has elsewhere or in the symbol table.

I get that, but I was hoping to be doing some good as this should support the other PRs that try to maintain types. Wouldn’t those suffer from the same issue? I’d happily use base_type_eq in the POSTCONDITION. (This PR is factored out from #2064, where the simplifier ended up returning completely unrelated types. I’d be happy to enforce semantic equivalence only.)

kroening commented:
@tautschnig I think the way to go is to get rid of base_type_eq and type_eq. This fixes the root cause of the issue.

kroening commented:
I.e., in the code in the PR, I'd replace the call to base_type_eq by expr.type() == expr.op().type().

There are a few places left where struct_tags get expanded (say the Java object factory). These should be fixed first.

tautschnig committed:
Simplify concatenation only if types match

When simplifying the operands of a concatenation permits replacing the concat
expression by one of the (simplified) operands, only do so if that wouldn't
result in a type mismatch.

tautschnig committed:
Expression simplification must not alter the type

Expression simplification should return an expression that is semantically
equivalent. The type may only differ in moving from a tag to an expanded type or
vice-versa, but the underlying types must always remain consistent.

tautschnig commented:
I have added some commits that remove unnecessary expansion of types, but 1) the Java front-end still needs work, 2) this PR is getting too big so I've started factoring out individual commits, and 3) I've borrowed commits from other PRs.

tautschnig commented:
I have cleared out all the dependencies, this PR is now really just adding extra checks in the simplifier while not going as far as what's going to happen in #2064 (more work in the string solver is required as it currently makes use of some elaborate expressions within types).

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 450e56850ba45382cb213d9b13fd1e78d3c4466a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98386718
